From: Richard Acayan <mailingradian@gmail.com>
Date: Tue, 8 Oct 2024 19:25:59 -0400
Subject: [PATCH] ASoC: qdsp6: q6voice: pass new topology property to q6cvp

The topology ID is now expected by q6cvp. Add the topology ID properties
to the paths with accessors, and pass them to the DSP.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>
---
 sound/soc/qcom/qdsp6/q6voice.c | 46 ++++++++++++++++++++++++++++++++++++------
 sound/soc/qcom/qdsp6/q6voice.h |  5 +++++
 2 files changed, 45 insertions(+), 6 deletions(-)

diff --git a/sound/soc/qcom/qdsp6/q6voice.c b/sound/soc/qcom/qdsp6/q6voice.c
index 8fd5c9c..461a5ec 100644
--- a/sound/soc/qcom/qdsp6/q6voice.c
+++ b/sound/soc/qcom/qdsp6/q6voice.c
@@ -13,6 +13,9 @@
 #include "q6mvm.h"
 #include "q6voice-common.h"
 
+#define VSS_IVOCPROC_TOPOLOGY_ID_TX_SM_ECNS		0x00010F71
+#define VSS_IVOCPROC_TOPOLOGY_ID_RX_DEFAULT		0x00010F77
+
 struct q6voice_path_runtime {
 	struct q6voice_session *sessions[Q6VOICE_SERVICE_COUNT];
 	unsigned int started;
@@ -23,6 +26,7 @@ struct q6voice_path {
 
 	enum q6voice_path_type type;
 	int tx_port, rx_port;
+	u32 tx_topo, rx_topo;
 	/* Serialize access to voice path session */
 	struct mutex lock;
 	struct q6voice_path_runtime *runtime;
@@ -35,17 +39,19 @@ struct q6voice {
 };
 
 static inline struct q6voice_session *session_create(enum q6voice_path_type type,
-						     int tx_port, int rx_port)
+						     int tx_port, int rx_port,
+						     u32 tx_topo, u32 rx_topo)
 {
-	return q6cvp_session_create(type, tx_port, rx_port);
+	return q6cvp_session_create(type, tx_port, rx_port, tx_topo, rx_topo);
 }
 
 static struct q6voice_session *session_create_v3(enum q6voice_path_type type,
-						 int tx_port, int rx_port)
+						 int tx_port, int rx_port,
+						 u32 tx_topo, u32 rx_topo)
 {
 	struct q6voice_session *cvp;
 
-	cvp = q6cvp_session_create_v3(type, tx_port, rx_port);
+	cvp = q6cvp_session_create_v3(type, tx_port, rx_port, tx_topo, rx_topo);
 	if (cvp == NULL)
 		return NULL;
 
@@ -81,11 +87,13 @@ static int q6voice_path_start(struct q6voice_path *p)
 		if (p->v->cvd_v2_3)
 			cvp = session_create_v3(p->type,
 						q6afe_get_port_id(p->tx_port),
-						q6afe_get_port_id(p->rx_port));
+						q6afe_get_port_id(p->rx_port),
+						p->tx_topo, p->rx_topo);
 		else
 			cvp = session_create(p->type,
 					     q6afe_get_port_id(p->tx_port),
-					     q6afe_get_port_id(p->rx_port));
+					     q6afe_get_port_id(p->rx_port),
+					     p->tx_topo, p->rx_topo);
 
 		if (IS_ERR(cvp))
 			return PTR_ERR(cvp);
@@ -259,6 +267,8 @@ struct q6voice *q6voice_create(struct device *dev, bool cvd_v2_3)
 
 		p->v = v;
 		p->type = path;
+		p->tx_topo = VSS_IVOCPROC_TOPOLOGY_ID_TX_SM_ECNS;
+		p->rx_topo = VSS_IVOCPROC_TOPOLOGY_ID_RX_DEFAULT;
 		mutex_init(&p->lock);
 	}
 
@@ -294,6 +304,30 @@ void q6voice_set_port(struct q6voice *v, enum q6voice_path_type path,
 }
 EXPORT_SYMBOL_GPL(q6voice_set_port);
 
+u32 q6voice_get_topology(struct q6voice *v, enum q6voice_path_type path,
+			 bool capture)
+{
+	struct q6voice_path *p = &v->paths[path];
+
+	if (capture)
+		return p->tx_topo;
+	else
+		return p->rx_topo;
+}
+EXPORT_SYMBOL_GPL(q6voice_get_topology);
+
+void q6voice_set_topology(struct q6voice *v, enum q6voice_path_type path,
+			  bool capture, u32 topo)
+{
+	struct q6voice_path *p = &v->paths[path];
+
+	if (capture)
+		p->tx_topo = topo;
+	else
+		p->rx_topo = topo;
+}
+EXPORT_SYMBOL_GPL(q6voice_set_topology);
+
 MODULE_AUTHOR("Stephan Gerhold <stephan@gerhold.net>");
 MODULE_DESCRIPTION("Q6Voice driver");
 MODULE_LICENSE("GPL v2");
diff --git a/sound/soc/qcom/qdsp6/q6voice.h b/sound/soc/qcom/qdsp6/q6voice.h
index a1e3300..3d002a9 100644
--- a/sound/soc/qcom/qdsp6/q6voice.h
+++ b/sound/soc/qcom/qdsp6/q6voice.h
@@ -25,4 +25,9 @@ int q6voice_get_port(struct q6voice *v, enum q6voice_path_type path,
 void q6voice_set_port(struct q6voice *v, enum q6voice_path_type path,
 		      bool capture, int index);
 
+u32 q6voice_get_topology(struct q6voice *v, enum q6voice_path_type path,
+			 bool capture);
+void q6voice_set_topology(struct q6voice *v, enum q6voice_path_type path,
+			  bool capture, u32 topo);
+
 #endif /*_Q6_VOICE_H */
